rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeper.java

```diff
@@ -208,14 +212,56 @@ public void setAllCached(String remotePath, long fileLength, long lastModified,
     public Map getCacheStats()
     {
         Map<String, Double> stats = new HashMap<String, Double>();
-        stats.put("Cache Hit Rate", ((double) cachedRequests / totalRequests));
-        stats.put("Cache Miss Rate", ((double) (totalRequests - cachedRequests) / totalRequests));
+        stats.put("Cache Hit Rate", ((double) cachedRequests / (cachedRequests + remoteRequests)));
+        stats.put("Cache Miss Rate", ((double) (remoteRequests) / (cachedRequests + remoteRequests)));
         stats.put("Cache Reads", ((double) cachedRequests));
         stats.put("Remote Reads", ((double) remoteRequests));
         stats.put("Non-Local Reads", ((double) (totalRequests - cachedRequests - remoteRequests)));
         return stats;
     }
 
+    //This method is to ensure that data required by another node is cached before it is read by that node
+    //using localTransferServer.
+    // If the data is not already cached, remoteReadRequest for that block is sent and data is cached.
+
+    //TODO : buffer is initialized everytime readData is called and it contains garbage value which is not required. Also, getCacheStatus is called twice.
+    @Override
+    public boolean readData(String path, long offset, int length, long fileSize, long lastModified, int clusterType)
+            throws TException
+    {
+        int blockSize = CacheConfig.getBlockSize(conf);
+        byte[] buffer = new byte[blockSize];
+        BookKeeperFactory bookKeeperFactory = new BookKeeperFactory(this);
+        CachingNativeS3FileSystem fs = null;
+        FSDataInputStream inputStream = null;
+        long startBlock = offset / blockSize;
+        long endBlock = ((offset + (length - 1)) / CacheConfig.getBlockSize(conf)) + 1;
+        try {
+            int idx = 0;
+            List<BlockLocation> blockLocations = getCacheStatus(path, fileSize, lastModified, startBlock, endBlock, clusterType);
+
+            for (int blockNum = (int) startBlock; blockNum < endBlock; blockNum++, idx++) {
+                int readStart = blockNum * blockSize;
+                if (blockLocations.get(idx).getLocation() != Location.CACHED) {
+                    if (fs == null) {
+                        fs = new CachingNativeS3FileSystem(bookKeeperFactory, new Path(path), conf);
+                        inputStream = fs.open(new Path(path), blockSize);
+                    }
+                    inputStream.seek(readStart);
+                    inputStream.read(buffer, 0, blockSize);
+                }
+            }
+            if (inputStream != null) {
+                inputStream.close();
+            }
+            return true;
+        }
+        catch (IOException e) {
+            log.info("Could not cache data: " + Throwables.getStackTraceAsString(e));
```

Comment:
close inputStream in the catch too.
Better add a finalize method and close stream there for all cases